src/query/src/part_sort.rs

```diff
@@ -418,15 +481,64 @@ impl PartSortStream {
             )
         })?;
 
-        // only clear after sorted for better debugging
-        self.buffer.clear();
         self.produced += sorted.num_rows();
         drop(full_input);
         // here remove both buffer and full_input memory
         self.reservation.shrink(2 * total_mem);
         Ok(sorted)
     }
 
+    /// Internal method for sorting `Top` buffer (with limit).
+    fn sort_top_buffer(&mut self) -> datafusion_common::Result<DfRecordBatch> {
+        let new_top_buffer = TopK::try_new(
+            self.partition,
+            self.schema().clone(),
+            vec![self.expression.clone()],
+            self.limit.unwrap(),
+            self.context.session_config().batch_size(),
+            self.context.runtime_env(),
+            &self.root_metrics,
+            self.partition,
+        )?;
+        let PartSortBuffer::Top(top_k, _) =
```

Comment:
Hmmm. It seems the row count number is useless.

Reply:
It's used to check if the buffer is empty. I've updated the doc comment